File: algorithmic_toolbox/1st_week/MaxPairProduct/maxpairproduct.cpp
#include <iostream>
#include <cstdlib>
#include <ctime>
#include <vector>
using namespace std;

long long maxpairproduct(const vector<long long> &A)
{
    long long result = 0;
    int n = A.size();

    for (int i = 0; i < n; i++)
    {
        for (int j = i + 1; j < n; j++)
        {
            if (A[i] * A[j] > result)
                result = A[i] * A[j];
        }
    }

    return result;
}

long long maxpairproductfast(const vector<long long> &A)
{
    int n = A.size();

    int index_max1 = -1;
    int index_max2 = -1;

    for (int i = 0; i < n; i++)
    {
        if (index_max1 == -1 || A[i] > A[index_max1])
            index_max1 = i;
    }

    for (int i = 0; i < n; i++)
    {
        if (i != index_max1 && (index_max2 == -1 || A[i] > A[index_max2]))
            index_max2 = i;
    }

    return A[index_max1] * A[index_max2];
}

int main()
{
    int n;
    cin >> n;

    vector<long long> A(n);
    for (int i = 0; i < n; i++)
        cin >> A[i];

    long long result = maxpairproductfast(A);

    cout << result << endl;

    return 0;
}

File: algorithmic_toolbox/1st_week/MaxPairProduct/maxpairproduct.py
def maxpairproduct(A):
    n = len(A)
    result = 0
    for f_index in range(n):
        for s_index in range(f_index + 1, n):
            if A[f_index] * A[s_index] > result:
                result = A[f_index] * A[s_index]
    return result


def maxpairproductfast(A):
    n = len(A)
    f_max = -1
    s_max = -1

    for index in range(n):
        if (f_max == -1) or (A[index] > A[f_max]):
            f_max = index

    for index in range(n):
        if (index != f_max) and ((s_max == -1) or (A[index] > A[s_max])):
            s_max = index

    return A[f_max] * A[s_max]


n = int(input())

A = [int(i) for i in input().split()]
print(maxpairproductfast(A))
